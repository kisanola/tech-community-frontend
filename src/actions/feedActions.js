import server from '../Api/server';

import * as types from '../actionTypes/feed';

import { FEED_LIMIT } from '../constants';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOOGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const setFeed = (payload) => ({
  type: types.SET_FEED,
  payload,
});

/**
 * Set feed organization
 * @return {object} action
 */
export const setFeedOrganizations = (payload) => ({
  type: types.SET_FEED_ORGANIZATIONS,
  payload,
});

/**
 * Get feed
 * @return {object} response
 */
export const getFeed = ({
  limit,
  itemsLength,
  category = 'general',
  search = '',
}) => async (dispatch) => {
  if (limit > itemsLength + FEED_LIMIT) return;

  dispatch(toggleLoading(true));

  try {
    const {
      data: { feed },
    } = await server.get(
      `/api/v1/feed?offset=${itemsLength}&limit=${FEED_LIMIT}&category=${category}&search=${search}`,
    );

    dispatch(setFeed(feed));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};


/**
 * Get feed
 * @return {object} response
 */
export const getFeedOrganizations = () => async (dispatch) => {
  const res = await server.get(
    '/api/v1/organizations?limit=5',
  );
  if (res) {
    const {
      data: { organizations },
    } = res;
    dispatch(setFeedOrganizations(organizations.docs || []));
  }
};
