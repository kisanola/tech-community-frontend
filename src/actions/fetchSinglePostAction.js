import server from '../Api/server';

import * as types from '../actionTypes/singlePostTypes';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOGGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const fetchSuccess = (post) => ({
  type: types.FETCH_SUCCESS,
  payload: post,
});

/**
 * Get feed
 * @return {object} response
 */
export default (slug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    const {
      data: { post },
    } = await server.get(
      `/posts/${slug}`,
    );

    dispatch(fetchSuccess(post));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
