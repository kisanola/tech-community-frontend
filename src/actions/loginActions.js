import axios from 'axios';
import {
  LOGIN_ERROR,
  LOGIN_SUCCESS,
  LOGIN_STARTED,
  LOGIN_INPUT_CHANGED,
} from '../actionTypes/loginTypes';

/**
 * Updates the form inputs
 * @param {*} attribute
 * @return {object} action
 */
export const formInputChanged = attribute => ({
  type: LOGIN_INPUT_CHANGED,
  payload: attribute,
});

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
export const loginStarted = () => ({
  type: LOGIN_STARTED,
});

/**
 * Triggers an action
 * when the process succeed
 * @param {object} user
 * @return {object} action
 */
export const loginSuccess = user => ({
  type: LOGIN_SUCCESS,
  payload: user,
});

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */

export const loginError = error => ({
  type: LOGIN_ERROR,
  payload: error,
});

/**
 * Login the user
 * @param {*} { username, email, password }
 * @return {object} response
 */
export const login = ({ email, password }) => async dispatch => {
  dispatch(loginStarted());

  try {
    const res = await axios.post(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/login`,
      {
        username: email,
        password,
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const { token, user } = res.data;

    localStorage.setItem('token', token);

    dispatch(loginSuccess(user));
    return true;
  } catch (e) {
    if (e.response) {
      if (e.response.data.errors) {
        dispatch(loginError(e.response.data.errors[0].message));
        return false;
      }
      dispatch(loginError(e.response.data.message));
      return false;
    }
    dispatch(loginError('Please check your internet connection'));
    return false;
  }
};
