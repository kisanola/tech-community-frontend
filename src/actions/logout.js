import { LOGOUT_SUCCESS } from '../actionTypes/currentUserTypes';

export const setIsLoggedOut = () => dispatch => {
  localStorage.removeItem('username');
  localStorage.removeItem('token');
  return dispatch({
    type: LOGOUT_SUCCESS,
    payload: true,
  });
};
