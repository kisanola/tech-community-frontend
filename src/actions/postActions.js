import axios from 'axios';
import server from '../Api/server';
import {
    POST_ERROR,
    POST_STARTED,
    POST_SUCCESS
} from '../actionTypes/postTypes';

const {
    REACT_APP_CLOUDINARY_URL_IMAGE,
    REACT_APP_CLOUDINARY_PRESET,
} = process.env;

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
const postStarted = () => ({ type: POST_STARTED })

/**
 * Triggers an action
 * when the process succeed
 * @param {object} post
 * @return {object} action
 */
const postSuccess = (post) => ({
    type: POST_SUCCESS,
    payload: post
})

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */
const postError = (error) => ({
    type: POST_ERROR,
    payload: error
})

const uploadImage = async file => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', REACT_APP_CLOUDINARY_PRESET);

    try {
        const res = await axios({
            url: REACT_APP_CLOUDINARY_URL_IMAGE,
            method: 'POST',
            header: {
                'Content-Type': 'application/x-ww-form-urlencoded',
            },
            data: formData,
        })
        return res.data.secure_url

    } catch (error) {
        throw error;
    }
};


/**
 * post
 * @param {*} { username, email, password }
 * @return {object} response
 */
const post = (data) => async dispatch => {
    dispatch(postStarted());

    try {
        const payload = {
            description: data.value,
            title: 'this is a wrong title'
        };

        if (data.image) payload.image = await uploadImage(data.image);

        const res = await server.post(`/api/v1/posts`, payload);

        const { post } = res.data;
        dispatch(postSuccess(post));
        return true;
    } catch (e) {
        if (e.response) {
            if (e.response.data.errors) {
                dispatch(postError('The post should have at least 50 characters long'));
                return false;
            }
            dispatch(postError(e.response.data.message));
            return false;
        }
        dispatch(postError('Please check your internet connection'));
        return false;
    }
};

export default post;

