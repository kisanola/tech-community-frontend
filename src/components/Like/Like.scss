@import '../../styles/colors';

.action {
  color: #888;

  i {
    font-size: 0.85em;
  }
  .count {
    font-size: 0.95em;
    margin-left: 10px;
  }

  &:hover {
    color: $primary;
    cursor: pointer;
  }
}

.liked {
  color: $primary;
  cursor: pointer;
}
