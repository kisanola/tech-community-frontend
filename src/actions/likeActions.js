import server from '../Api/server';

import { toggleLoading } from './feedActions';

/**
 * Like a post
 * @return {object} action
 */
export const like = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.post(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};

/**
 * Unlike a post
 * @return {object} action
 */
export const unlike = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.delete(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
