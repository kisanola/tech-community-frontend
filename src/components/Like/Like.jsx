import React, { Component } from 'react';
import { connect } from 'react-redux';
import './Like.scss';
import { like, unlike } from '../../actions/likeActions';
import { handleShowAndHide as toggleSocialModal } from '../../actions/socialAuth';

export class Like extends Component {
  state = {
    highlight: false,
    likesCount: 0,
  };

  componentWillMount() {
    const { liked = false, likesCount } = this.props;
    this.setState({ highlight: liked, likesCount });
  }

  toggleLikeHighlight = () => {
    const { likesCount, highlight } = this.state;
    const checkLimit = () =>
      highlight && likesCount > 0 ? likesCount - 1 : likesCount + 1;

    this.setState({ highlight: !highlight, likesCount: checkLimit() });
  };

  handleLike = () => {
    const { highlight } = this.state;
    const { slug, onLike, onUnlike, isAuth, onToggleSocialModal } = this.props;

    if (!isAuth) return onToggleSocialModal(true);

    this.toggleLikeHighlight();

    if (highlight) return onUnlike(slug);

    onLike(slug);
  };

  render() {
    const {
      handleLike,
      state: { highlight, likesCount },
    } = this;

    return (
      <div
        className={`action ${highlight && 'liked'}`}
        title="Like"
        onClick={() => handleLike()}
      >
        <i className="fas fa-thumbs-up" />
        <span className="count">{likesCount}</span>
      </div>
    );
  }
}

export const mapStateToProps = ({ currentUser: { isAuth } }) => ({
  isAuth,
});

export const mapDispatchToProps = dispatch => ({
  onLike: slug => dispatch(like(slug)),
  onUnlike: slug => dispatch(unlike(slug)),
  onToggleSocialModal: show => dispatch(toggleSocialModal(show)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Like);
