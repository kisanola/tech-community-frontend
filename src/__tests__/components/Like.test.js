import React from 'react';
import { shallow, mount } from 'enzyme';
import Like, {
  Like as LikeComponent,
  mapStateToProps,
  mapDispatchToProps,
} from '../../components/Like/Like';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';
import moxios from 'moxios';
import { like, unlike } from '../../actions/likeActions';

describe('Like', () => {
  let store;
  let axiosInstance;
  const mockStore = configureMockStore([thunk]);
  store = mockStore({});

  let wrapper;
  const props = {
    isAuth: false,
    onLike: jest.fn(),
    onUnlike: jest.fn(),
    onToggleSocialModal: jest.fn(),
    slug: 'article-1234',
    likesCount: 2,
  };

  beforeAll(() => {
    wrapper = shallow(
      <Provider store={store}>
        <Like />
      </Provider>,
    );
  });

  describe('Like.jsx', () => {
    it('should render the component', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('`LikeComponent`', () => {
    it('should call `onToggleSocialModal`', () => {
      wrapper = mount(<LikeComponent {...props} />);
      wrapper.find('.action').simulate('click');
      expect(wrapper.props().onToggleSocialModal).toHaveBeenCalled();
    });

    it('should call `onLike`', () => {
      wrapper = mount(<LikeComponent {...{ ...props, isAuth: true }} />);
      wrapper.find('.action').simulate('click');
      expect(wrapper.props().onLike).toHaveBeenCalled();
    });

    it('should call `onUnlike`', () => {
      wrapper = mount(
        <LikeComponent {...{ ...props, liked: true, isAuth: true }} />,
      );
      wrapper.find('.action').simulate('click');
      expect(wrapper.props().onUnlike).toHaveBeenCalled();
    });
  });

  describe('State', () => {
    let mapped, dispatch;
    beforeEach(() => {
      dispatch = jest.fn();
      mapped = mapDispatchToProps(dispatch);
    });

    test('should return `mapStateToProps`', () => {
      const expectedState = {
        isAuth: false,
      };
      const state = mapStateToProps({
        currentUser: { isAuth: false },
      });
      expect(state).toEqual(expectedState);
    });

    it('Should call onLike', () => {
      mapped.onLike('slug');
      expect(dispatch).toHaveBeenCalled();
    });

    it('Should call onUnlike', () => {
      mapped.onUnlike('slug');
      expect(dispatch).toHaveBeenCalled();
    });

    it('Should call onToggleSocialModal', () => {
      mapped.onToggleSocialModal(true);
      expect(dispatch).toHaveBeenCalled();
    });
  });

  describe('Login actions', () => {
    beforeEach(() => {
      store = mockStore({});
      axiosInstance = axios;
      moxios.install(axiosInstance);
    });

    afterEach(() => {
      moxios.uninstall(axios);
    });

    test('Should call the `like` action with success', () => {
      moxios.stubRequest(
        `${process.env.REACT_APP_BACKEND_URL}/api/v1/post/slug`,
        {
          status: 200,
          response: {},
        },
      );

      store.dispatch(like('slug')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('Should call the `like` action with error', () => {
      moxios.stubRequest(
        `${process.env.REACT_APP_BACKEND_URL}/api/v1/post/slug`,
        {
          status: 409,
          response: {},
        },
      );

      store.dispatch(like('slug')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('Should call the `unLike` action with success', () => {
      moxios.stubRequest(
        `${process.env.REACT_APP_BACKEND_URL}/api/v1/post/slug`,
        {
          status: 200,
          response: {},
        },
      );

      store.dispatch(unlike('slug')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    test('Should call the `unLike` action with error', () => {
      moxios.stubRequest(
        `${process.env.REACT_APP_BACKEND_URL}/api/v1/post/slug`,
        {
          status: 409,
          response: {},
        },
      );

      store.dispatch(unlike('slug')).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });
});
