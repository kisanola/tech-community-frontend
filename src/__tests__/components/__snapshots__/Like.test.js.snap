// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Like Like.jsx should render the component 1`] = `ShallowWrapper {}`;

exports[`Like Login actions Should call the \`like\` action with error 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;

exports[`Like Login actions Should call the \`like\` action with success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;

exports[`Like Login actions Should call the \`unLike\` action with error 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;

exports[`Like Login actions Should call the \`unLike\` action with success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;
